src/sentry/api/serializers/models/rule.py

```diff
@@ -64,25 +67,20 @@ def get_attrs(self, item_list, user, **kwargs):
         resolved_actors = {}
         owners_by_type = defaultdict(list)
 
-        sentry_app_uuids = {
+        sentry_app_uuids = [
             action.get("sentryAppInstallationUuid")
             for rule in rules.values()
             for action in rule.data.get("actions", [])
-        }
-
-        sentry_app_ids = (
-            SentryAppInstallation.objects.filter(uuid__in=sentry_app_uuids)
-            .distinct("sentry_app_id")
-            .values_list("sentry_app_id", flat=True)
-        )
+        ]
 
-        sentry_app_installations_by_uuid = (
-            SentryAppInstallation.objects.get_related_sentry_app_components(
-                organization_ids={rule.project.organization_id for rule in rules.values()},
-                sentry_app_ids=sentry_app_ids,
-                type="alert-rule-action",
-                group_by="uuid",
-            )
+        sentry_app_ids: List[int] = [
+            i.sentry_app.id for i in app_service.get_many(filter=dict(uuids=sentry_app_uuids))
+        ]
+        sentry_app_installations_by_uuid = app_service.get_related_sentry_app_components(
+            organization_ids=[rule.project.organization_id for rule in rules.values()],
```

Comment:
I hope all the rules have the project eager loaded 😬

Reply:
This is one of the problems with our serializer architecture -- it's not clear because this is just the "item_list", so it is the caller's responsibility to handle this.  I can't alter the item list to enforce the preload because it is often just a list and not a queryset.

Reply:
Ultimately, unpopular opinion here, but serializers should likely only operate on preloaded, static, dumb data; classes or dictionaries.  It should have one job: deterministically reshape a payload.  

Controllers are the kinds of things that should carry work related to querying, relations, and normalization.

Reply:
And if it needs to be shared, it should just be a dumb function somewhere that preloads into a dumb object, with as few other abstractions as possible.

Reply:
The general point of `get_attrs` is to perform bulk load operations. There are lots of legacy cases where the caller is expected to preload data, but really the correct way is to bulk load all data in getattrs, and assume it hasn't been done.

`prefetch_related_objects` lets you do this with a list of models fairly easily.

Reply:
Interesting, thanks for the tip on `prefetch_related_objects`, I'll try adding that.